perception/autoware_tensorrt_bevdet/src/bevdet.cpp

```diff
@@ -0,0 +1,717 @@
+// Copyright 2024 AutoCore, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// cspell:ignore BEVDET, thre, TRTBEV, bevdet, caminfo, intrin, Ncams, bevfeat, bevpool, maxnum,
+// zgrid cspell:ignore gridbev, egobev, adjgrid, currgrid
+#include "autoware/tensorrt_bevdet/bevdet.hpp"
+
+#include "autoware/tensorrt_bevdet/alignbev_plugin.hpp"
+#include "autoware/tensorrt_bevdet/bevpool_plugin.hpp"
+#include "autoware/tensorrt_bevdet/gatherbev_plugin.hpp"
+#include "autoware/tensorrt_bevdet/preprocess_plugin.hpp"
+
+#include <thrust/functional.h>
+#include <thrust/sort.h>
+
+#include <chrono>
+#include <cstdio>
+#include <fstream>
+#include <iostream>
+
+using std::chrono::duration;
+using std::chrono::high_resolution_clock;
+
+namespace autoware::tensorrt_bevdet
+{
+BEVDet::BEVDet(
+  const std::string & config_file, int n_img, std::vector<Eigen::Matrix3f> cams_intrin,
+  std::vector<Eigen::Quaternion<float>> cams2ego_rot,
+  std::vector<Eigen::Translation3f> cams2ego_trans, const std::string & onnx_file,
+  const std::string & engine_file)
+: cams_intrin_(cams_intrin), cams2ego_rot_(cams2ego_rot), cams2ego_trans_(cams2ego_trans)
+{
+  initParams(config_file);
+  if (n_img != n_img_) {
+    RCLCPP_ERROR(
+      rclcpp::get_logger("BEVDet"), "BEVDet need %d images, but given %d images!", n_img_, n_img);
+  }
+  auto start = high_resolution_clock::now();
+
+  std::shared_ptr<int> ranks_bev_ptr = nullptr;
```

Comment:
Consider using `std::unique_ptr` instead of raw pointers for `ranks_bev_ptr`, `ranks_depth_ptr`, `ranks_feat_ptr`, `interval_starts_ptr`, and `interval_lengths_ptr`. This will improve memory management and prevent potential memory leaks.